base/iterators.jl

```diff
@@ -648,63 +595,115 @@ julia> collect(Iterators.product(1:2,3:5))
-iteratorsize(::Type{Prod2{I1,I2}}) where {I1,I2} = prod_iteratorsize(iteratorsize(I1),iteratorsize(I2))
+prod_iteratorsize(::Union{HasLength,HasShape}, ::Union{HasLength,HasShape}) = HasShape()
+# products can have an infinite iterator
+prod_iteratorsize(::IsInfinite, ::IsInfinite) = IsInfinite()
+prod_iteratorsize(a, ::IsInfinite) = IsInfinite()
+prod_iteratorsize(::IsInfinite, b) = IsInfinite()
+prod_iteratorsize(a, b) = SizeUnknown()
 
-function start(p::AbstractProdIterator)
-    s1, s2 = start(p.a), start(p.b)
-    s1, s2, Nullable{eltype(p.b)}(), (done(p.a,s1) || done(p.b,s2))
-end
+size(P::ProductIterator) = _prod_size(P.iterators)
+_prod_size(::Tuple{}) = ()
+_prod_size(t::Tuple) = tuple(_prod_size1(t[1], iteratorsize(t[1]))..., _prod_size(tail(t))...)
+_prod_size1(a, ::HasShape)  = size(a)
+_prod_size1(a, ::HasLength) = (length(a),)
+_prod_size1(a, A) =
+    throw(ArgumentError("Cannot compute size for object of type $(typeof(a))"))
 
-@inline function prod_next(p, st)
-    s1, s2 = st[1], st[2]
-    v1, s1 = next(p.a, s1)
+indices(P::ProductIterator) = _prod_indices(P.iterators)
+_prod_indices(::Tuple{}) = ()
+_prod_indices(t::Tuple) = tuple(_prod_indices1(t[1], iteratorsize(t[1]))..., _prod_indices(tail(t))...)
+_prod_indices1(a, ::HasShape)  = indices(a)
+_prod_indices1(a, ::HasLength) = (OneTo(length(a)),)
+_prod_indices1(a, A) =
+    throw(ArgumentError("Cannot compute indices for object of type $(typeof(a))"))
 
-    nv2 = st[3]
-    if isnull(nv2)
-        v2, s2 = next(p.b, s2)
-    else
-        v2 = nv2.value
+ndims(p::ProductIterator) = length(indices(p))
+length(P::ProductIterator) = prod(size(P))
+_length(p::ProductIterator) = prod(map(unsafe_length, indices(p)))
+
+iteratoreltype(::Type{ProductIterator{Tuple{}}}) = HasEltype()
+iteratoreltype(::Type{ProductIterator{Tuple{I}}}) where {I} = iteratoreltype(I)
+function iteratoreltype(::Type{ProductIterator{T}}) where {T<:Tuple}
+    I = tuple_type_head(T)
+    P = ProductIterator{tuple_type_tail(T)}
+    iteratoreltype(I) == EltypeUnknown() ? EltypeUnknown() : iteratoreltype(P)
+end
+
+eltype(P::ProductIterator) = _prod_eltype(P.iterators)
+_prod_eltype(::Tuple{}) = Tuple{}
+_prod_eltype(t::Tuple) = Base.tuple_type_cons(eltype(t[1]),_prod_eltype(tail(t)))
+
+start(::ProductIterator{Tuple{}}) = false
+next(::ProductIterator{Tuple{}}, state) = (), true
+done(::ProductIterator{Tuple{}}, state) = state
+
+function start(P::ProductIterator)
+    iterators = P.iterators
+    iter1 = first(iterators)
+    state1 = start(iter1)
+    d, states, nvalues = _prod_start(tail(iterators))
+    d |= done(iter1, state1)
+    return (d, (state1, states...), nvalues)
+end
+function next(P::ProductIterator, state)
+    iterators = P.iterators
+    d, states, nvalues = state
+    iter1 = first(iterators)
+    value1, state1 = next(iter1, states[1])
+    tailstates = tail(states)
+    values = (value1, map(unsafe_get, state[3])...) # safe if not done(P, state)
```

Comment:
`state[3]` should be equal to `nvalues` here. However, I came across a bug where, if I write `nvalues` here, it seems like it is taking the "value" of  `nvalues`  after it has been overwritten in the next block of code. In particular, in the final state before being done, none of the elements in `nvalues` will be `isnull`, but then, after the next block of code, they are. 

Put differently, this function produced a different answer when I just call it versus when I run it line by line in the REPL. Or yet differently, if you replace `state[3]` by `nvalues` in this place, which should be a valid replacement, the tests fail outside of an actual test. I will analyse further and file a more complete bug description if I am convinced of my case.